index.md

```diff
@@ -1,16 +1,128 @@
+5. The payload is delivered as an installation script.  The DM pipelines stack can also be installed this way with [lsstinstall][2] and this may be the most common way for thorny environments to be delivered.  For anything of sufficent size, that has made the (bad) choice to not use a standard build/installation system, this will probably already exist.  See, for instance, the huge proliferation in products that expect you to `curl https://some-url | sudo /bin/bash`.  Less grotesquely, maybe it's something you can install with `./configure; make; make install`.
+
+My preference is the second method.  That lets us keep almost all of the image-building machinery the same between payloads, and by encapsulating the payload installation in a script and image layer of its own, replacement of the payload becomes quite straightforward.  If the payload is in one of the easy formats, this script might literally be nothing more than `pip install -r requirements.txt`.  However, specific image-building considerations mean that it may not be a single script: system-level packages may need to be installed as root, while we would prefer that both the installation and runtime operation of the payload happen as a less-privileged user.
+
+That sets up the third point: what will be necessary in order to turn the installed payload into a Jupyterlab-runnable kernel and a terminal environment with the payload contents configured and available?
+
+## Payload configuration
+
+Configuration of the payload environment has at least three components.
+
+Setting up the Jupyter kernel is the most straightforward.  If the payload can be represented in a Python virtual environment, `python -m ipykernel install` is all that needs to be done.  If the payload is more complicated (as the DM pipelines stack is), then it turns out that the kernel can be a shell script that does necessary setup and execs a python process as its last action.
+
+The next is setting up a terminal environment.  One way to do this is to set a sentinel environment variable at Jupyterlab launch, and then when the user requests a shell, in the user's `.bashrc` or moral equivalent, test that variable, and if it is set (indicating the user is running inside Jupyterlab), perform the necessary setup (environment activation, running an appropriate setup script, whatever) before yielding control to the user for interactive shell use.
+
+The third is the most difficult.  In Nublado, we have a spawner form.  That form contains a selection of prepulled images, a drop-down box to select an uncached image, a selector to choose container size, and two checkboxes useful for debugging and unwedging stuck environments.  I am not actually sure how generic that is, but I think erring on the side of "everyone will want something like this" is better than insisting it's DM-pipelines-specific.
```

Comment:
Minor correction: I think the debugging setting is for JupyterLab, not for the stack environment? Or at least it's for both. Doesn't it enable debug logging in JupyterLab? We probably want to keep that separate from any payload since it's independently helpful.

Reply:
I wrote "stuck" environment but I will choose a different adjective not so easily readable as "stack".  What I meant was, "User reports their container won't start; we need some debugging information to determine where it's choking."

Perhaps "inoperable".

Reply:
Oh, sorry, I indeed misread "stuck" as "stack."